### ISSUE TITLE
Add way of deploying code+engine+ extensions as a java executable

### ISSUE THREAD

**Issue by BruceEdmonds** [NetLogo/NetLogo#1308]
As extensions become more popular, there is a multiple Netlogo version + extensions versions problem. One not only has to find the right version of Netlogo to run a model, but also the right versions of the extensions. This can make it hard for others to look at and play with one's model.

How difficult would it be to package model+engine+listed_extensions all together into some kind of java package (bytecode or a .jar file?) that could then be distributed as a single file and run without too much difficulty?

**Comment by mrerrormessage** [comment]
What you're describing sounds somewhat similar to the NetLogo applets that could be made in NetLogo 5.x and earlier. These were somewhat self-contained packages which could be deployed without too much hassle and provided a working model with zero user configuration. Of course, Java Applets are no longer supported (either by NetLogo or by Oracle) and there hasn't been a full replacement for users who want to easily distribute a NetLogo Desktop model. 

None of this is to say that we haven't noticed the problem with extensions being a hassle to download. Note that it is possible to include extensions in the same directory as the model itself, which can make it easier to deploy models with their extensions packaged as a zip or tar file. Additionally, we've had a proposal for [an extension downloader](https://github.com/NetLogo/NetLogo/issues/736) on our radar for at least two years now. An extension downloader would be quite a bit simpler to build than a tool that packaged NetLogo, but would still require a substantial amount of effort, both in implementing within NetLogo and in aggregating and maintaining a list of extensions for NetLogo and their compatibility with different versions.

Similarly, for users who want to put the same model in front of a lot of different people (like classroom teachers) we've been working hard to make NetLogo Web a great solution for interacting with NetLogo models. NetLogo Web doesn't require java installation and can be run with just a browser, making it the best choice for cross-platform deployment of NetLogo models.

Regarding the question of difficulty, what you're requesting is "very difficult" for reasons I can elaborate on if you're interested. The specific difficulty depends on what you're actually requesting. If the request is merely to provide hooks within the NetLogo controlling API for this arrangement, that is actually pretty straightforward and could probably be accomplished in a reasonable amount of time. If the request is to provide something like a "save as Java executable" menu item, then the task is staggeringly difficult. In either case, the process of deploying NetLogo as an executable or single jar is quite difficult, in the first case that difficulty is shifted off of the NetLogo developers and onto controlling API authors whereas in the second case it falls entirely on the NetLogo developers.

**Comment by SethTisue** [comment]
> Note that it is possible to include extensions in the same directory as the model itself, which can make it easier to deploy models with their extensions packaged as a zip or tar file.

Yeah. That's the solution I would recommend for models that don't work in NetLogo Web.

I suspect many model authors aren't aware they can do this.

**Comment by SethTisue** [comment]
> Note that it is possible to include extensions in the same directory as the model itself, which can make it easier to deploy models with their extensions packaged as a zip or tar file.

Yeah. That's the solution I would recommend for models that don't work in NetLogo Web.

I suspect many model authors aren't aware they can do this.
